### PR TITLE
Mana sinks & faucets

### PR DESCRIPTION
**Pull request by iansui**
initially you have 100 mana
you get 10 mana per 10 seconds
each fireball gives you 2 mana
each big enemy gives you 20 mana
each small enemy gives you 10 mana
each big friendly unit costs 25 mana
each small friendly unit costs 15 mana

related issues: #50 #18

### CODE REVIEW

File: TowerNinja/Assets/Scripts/FriendSpawn.cs [hunk 1]
```diff
@@ -6,15 +6,26 @@
...
+    private static readonly int BigFriendManaCost = 25;
+    private static readonly int FriendManaCost = 15;
+    private bool spawningBigFriend;
+
     void Start()
     {
-        
+        // Get resource manager
+        GameObject resourceManagerObject = GameObject.Find("ResourceManager").gameObject;
+        resourceManager = resourceManagerObject.GetComponent<ResourceManager>();
```
Comment (eye942) [inline TowerNinja/Assets/Scripts/FriendSpawn.cs:22]: same as above comment

File: TowerNinja/Assets/Scripts/enemyDmgHitbox.cs [hunk 2]
```diff
@@ -114,12 +117,31 @@ public void DamageEnemy(int damage)
...
             //Debug.Log(this.gameObject + "is destroyed.");
             Destroy(this.gameObject);
             ReportEnemyDeath();
         }
     }
 
+    private void GiveManaReward()
+    {
+        GameObject resourceManagerObject = GameObject.Find("ResourceManager").gameObject;
+        ResourceManager resourceManager = resourceManagerObject.GetComponent<ResourceManager>();
```
Comment (eye942) [inline TowerNinja/Assets/Scripts/enemyDmgHitbox.cs:132]: Same as above

File: TowerNinja/Assets/Scripts/enemyDmgHitbox.cs [hunk 3]
```diff
@@ -114,12 +117,31 @@ public void DamageEnemy(int damage)
...
+
+        if (gameObject.name.StartsWith("Big"))
+        {
+            resourceManager.IncreaseMana(BigEnemyManaReward);
+        }
+        else
+        {
+            resourceManager.IncreaseMana(EnemyManaReward);
+        }
```
Comment (eye942) [inline TowerNinja/Assets/Scripts/enemyDmgHitbox.cs]: It's prolly best to set the EnemyManaReward from the unity via prefab.
Then we can just do 1 line

File: TowerNinja/Assets/Scripts/Fireball.cs [hunk 4]
```diff
@@ -104,6 +108,13 @@ private void Die()
         Destroy(gameObject);
     }
 
+    private void GiveManaReward()
+    {
+        GameObject resourceManagerObject = GameObject.Find("ResourceManager").gameObject;
+        ResourceManager resourceManager = resourceManagerObject.GetComponent<ResourceManager>();
```
Comment (eye942) [inline TowerNinja/Assets/Scripts/Fireball.cs:114]: You can simplify the above to
`var resourceManager = FindObjectOfType<ResourceManager>()`

File: TowerNinja/Assets/Scripts/FriendSpawn.cs [hunk 5]
```diff
@@ -6,15 +6,26 @@
...
-    
     public Transform prefab;
     public Vector3 velocity;
     public Transform friend;
     public GameObject go;
     public KeyCode activateKey;
+
+    private ResourceManager resourceManager;
+    private static readonly int BigFriendManaCost = 25;
+    private static readonly int FriendManaCost = 15;
```
Comment (eye942) [inline TowerNinja/Assets/Scripts/FriendSpawn.cs]: lets make 1 ManaCost variable and set it via Unity's prefab.

**Review by eye942 — CHANGES_REQUESTED** [review]

**Comment by iansui** [comment]
@eye942 
I looked it up and it seems like FindObjectOfType() is a lot slower than GameObject.Find() & GetComponent(),
and FindObjectOfType() is actually discouraged in general cases. Unity profilers will give a performance warning.
Because FindObjectOfType() does hierarchical recursive search on every gameobject in the scene
GameObject.Find("name") uses hashed string to locate the gameobject directly and GetComponent() only looks for components under that gameobject

Also can you give an example of "make it a variable and set it via Unity's prefab"?
I don't know how to set it up in the unity editor

**Comment by eye942** [comment]
You can create variable values by creating public field in a script.

You can then set them via script component in Unity editor
image

**Review by eye942 — APPROVED** [review]

**Pull request merged**